### PR TITLE
block-layer/codecs: dag-pb can't enforce PBNode ordering

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

### CODE REVIEW

**Comment by willscott** [comment]
i'm not sure i agree with the conclusion of not having a canonical form?
just because we're willing to decode non-canonical stored representations doesn't mean there isn't a canonical form

**Review by willscott — APPROVED** [review]

**Comment by mvdan** [comment]
I guess that depends on whether you define encoders that use a different order as spec-compliant or not. This is why I said the encoding "should" use a specific order, instead of "must".

If we don't get a review from Rod by the middle of next week, I can always remove that sentence to unblock a merge.

**Comment by rvagg** [comment]
Yeah, I'm with @willscott on this one.

I think it should be "**must** be encoded ...", but "The decoder should accept either order", and also that there **is** a canonical form for given data. It's just that there exists some historical data which a decoder may encounter which is not in that canonical form. What we want here is for any new/current codec implementation to produce that canonical form, so round-trips of this historical data will end up as different bytes & with a different CID. We're not committing to preserving non-canonical forms in round-trips either (which is one way you could read this current wording I think).

**Comment by mvdan** [comment]
Fair enough, that's a good point. Your wording does imply that someone using a "field number order" protobuf encoder is actively breaking the spec, but I think we're fine with that for the sake of strongly encouraging canonical forms.

**Comment by mvdan** [comment]
Updated :) I'll prepare a branch for go-codec-dagpb on top of the performance PR, since otherwise I think I'll run into conflicts with myself.

File: block-layer/codecs/dag-pb.md
Comment (warpfork) [inline block-layer/codecs/dag-pb.md]:
```suggestion
1. Fields in the `PBLink` message must appear in the order as defined by the Protobuf schema above, following the field numbers. Blocks with out-of-order `PBLink` fields should be rejected.  (Note that it is common for Protobuf decoders to accept out-of-order field entries, which means the DAG-PB spec is somewhat stricter than may be seen as typical for other Protobuf-based formats.)
```

Comment (warpfork) [reply block-layer/codecs/dag-pb.md]: I wonder if the "following the field numbers" part should be moved to a paranthetical or second sentence also, since I notice the line right below this has to make a "Note [...]" of the fact it *doesn't* follow the field numbers.

(But of course these are all nits on very fine details of prose, just striving for clarity but saying the same things already said.)

File: block-layer/codecs/dag-pb.md
Comment (warpfork) [inline block-layer/codecs/dag-pb.md]:
```suggestion
3. Duplicate entries in the binary form are invalid; blocks with duplicate field values should be rejected. (Note that it is common for Protobuf decoders to accept repeated field values in the binary data, and interpret them _updates_ to fields that have already been set; DAG-PB is stricter than this.)
```

(if I'm understanding protobuf correctly, that is)

File: block-layer/codecs/dag-pb.md
Comment (warpfork) [inline block-layer/codecs/dag-pb.md]: I'd kind of like to move the contextualizing commentary about protobuf norms to a paranthetical here, too, since it's more about protobuf norms than it is about DAG-PB.

**Review by warpfork — APPROVED** [review]
I'm really appreciative of the time spent to verify all these details and make sure our spec discusses them accurately and well!  :smile: 

---

Thinking-out-loud(-but-writing-it-down) for a bit, just for posterity:

- Most of the time (e.g., as I have with `*json` and `*cbor`), I think I'd argue that if the variation is common in the wild, and has any chance of being loadbearing, we should admit it and be able to round-trip it through the data model.
- ...But in this case, I think the former is sorta-true (but presumably there's a strong predominance to one ordering), and the latter we definitely know to be untrue (it's *not* allowed to be load-bearing (other than that it affects the hash)).

Resolutions which come from that:
- I have a harder time than usual imagining a practical call to have support for loading the non-canonical order and re-emitting that with our libraries.
- ... and therefore, yes, I feel comfortable with the strength of the wording proposed here about the necessity of encoding following the strict order.
- Basically +1 to what @rvagg  said.

---

I made a couple suggestions on the prose, mostly about clarifying some comments that are about protobuf norms, but the whole thing reads quite well already.  Nice work.

File: block-layer/codecs/dag-pb.md
Comment (mvdan) [reply block-layer/codecs/dag-pb.md]: ack, done

**Comment by mvdan** [comment]
I've applied all three of Eric's nits. It seems like we all agree, so I'm moving ahead. Thanks!

**Pull request merged**
